### PR TITLE
Fix suggestion when the argument is a hash. Fixes #267

### LINKED ISSUE #267: Bug in SingleArgumentMessageChain
If you (for some strange reason) are calling `receive_message_chain` with a hash, you should substitute it with `receive_messages` instead of just `receive`. E.g.:

```rb
allow(foo).to receive_message_chain(bar: 42, baz: 42)
```
is suggested by the cop to be replaced with
```rb
allow(foo).to receive(bar: 42, baz: 42)
```
while the actually working code would be
```rb
allow(foo).to receive_messages(bar: 42, baz: 42)
```

### PR DESCRIPTION
**Pull request by Darhazer**
@bquorning

### CODE REVIEW

**Comment by backus** [comment]
@Darhazer I'm not sure these are equivalent?

```
>> require 'rspec/mocks/standalone'
=> true
>> dbl = double(:foo)
=> #<Double :foo>
>> allow(dbl).to receive_message_chain(foo: 1, bar: 2)
=> []
>> dbl.foo
=> #<Double (anonymous)>
>> dbl.foo.bar
=> 2
>> dbl.bar
RSpec::Mocks::MockExpectationError: #<Double :foo> received unexpected message :bar with (no args)
from /Users/johnbackus/.gem/ruby/2.4.0/gems/rspec-support-3.6.0.beta2/lib/rspec/support.rb:87:in 'block in <module:Support>'
>> dbl2 = double(:foo)
=> #<Double :foo>
>> allow(dbl2).to receive_messages(foo: 1, bar: 2)
>> dbl2.foo
=> 1
>> dbl2.bar
=> 2
>> dbl2.foo.bar
NoMethodError: undefined method `bar' for 1:Integer
from (pry):13:in `__pry__'
```

**Comment by Darhazer** [comment]
Then #267 seems invalid

**Comment by backus** [comment]
Yeah. @bquorning can you confirm?

**Comment by bquorning** [comment]
I stand corrected, my suggestion in #267 was incorrect. But the cop is still incorrect in suggesting 
`receive` as a substitute for `receive_message_chain` in
```rb
allow(foo).to receive_message_chain(bar: 42, baz: 42)
```

When a singular argument to `receive_message_chain` is a hash, `RSpec/SingleArgumentMessageChain` should report no offense.

**Comment by Darhazer** [comment]
If that's the case, it's even simpler to fix and I updated the PR.

btw, what is the reason for `first.to_s.include?('.')` check?

**Comment by bquorning** [comment]
> btw, what is the reason for `first.to_s.include?('.')` check?

That’s to allow `allow(foo).to receive_message_chain("one.two") { :three }`. It’s not a syntax that I’d use, but I guess we have to support it.

**Comment by bquorning** [comment]
You should probably add an entry to the changelog.

**Comment by Darhazer** [comment]
Changelog entry added.

@bquorning should we handle the case when the argument is a hash with single key, like
`allow(foo).to receive_message_chain(bar: 42)`

**Comment by bquorning** [comment]
> should we handle the case when the argument is a hash with single key

That sounds like a good idea. These two are equivalent:
```rb
allow(foo).to receive_message_chain(bar: 42)
allow(foo).to receive(:bar) { 42 }
```

**Comment by dgollahon** [comment]
> That sounds like a good idea. These two are equivalent:
>
> ```
> allow(foo).to receive_message_chain(bar: 42)
> allow(foo).to receive(:bar) { 42 }
> ```

```ruby
allow(foo).to receive(:bar) { 42 }
```

is also equivalent to

```ruby
allow(foo).to receive(:bar).and_return(42)
```

which I think is more intention revealing / nicer if we're going to correct it.

For that matter, it'd probably make a decent cop (though correcting from `receive(foo) { bar }` to `receive(foo).and_return(bar)` is not always truly equivalent due to evaluation order).

**Comment by Darhazer** [comment]
PR updated

File: spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb [hunk 1]
```diff
@@ -47,6 +47,49 @@
...
+          end
+        RUBY
+      end
+    end
+
+    context 'with empty hash' do
+      it "doesn't report an offence" do
+        expect_no_violations(<<-RUBY)
+          before do
+            allow(foo).to receive_message_chain({})
```
Comment (bquorning) [inline spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb]: I’m not sure what `receive_message_chain({})` does. Intuitively, I’d think it doesn’t do anything.

File: spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb [hunk 2]
```diff
@@ -47,6 +47,49 @@
...
+      include_examples(
+        'autocorrect',
+        'before { allow(foo).to receive_message_chain(bar: 42) }',
+        'before { allow(foo).to receive(:bar) { 42 } }'
+      )
+
+      include_examples(
+        'autocorrect',
+        'before { allow(foo).to receive_message_chain("foo bar" => 42) }',
+        'before { allow(foo).to receive("foo bar") { 42 } }'
```
Comment (bquorning) [inline spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb]: Can we do that, with a string including whitespace?

Comment (Darhazer) [inline spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb]: lol, I was so concerned to handle all hash cases that I forgot they are used in a `receive` call. Guess all this safety is unnecessary

**Comment by Darhazer** [comment]
PR updated

File: spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb [hunk 3]
```diff
@@ -47,6 +47,39 @@
...
+      include_examples(
+        'autocorrect',
+        'before { allow(foo).to receive_message_chain(bar: 42) }',
+        'before { allow(foo).to receive(:bar) { 42 } }'
+      )
+
+      include_examples(
+        'autocorrect',
+        'before { allow(foo).to receive_message_chain("bar" => 42) }',
+        'before { allow(foo).to receive(:bar) { 42 } }'
```
Comment (dgollahon) [inline spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb]: I wonder if we should preserve the key type.

File: lib/rubocop/cop/rspec/single_argument_message_chain.rb [hunk 4]
```diff
@@ -22,22 +22,42 @@ class SingleArgumentMessageChain < Cop
           (send _ #{Matchers::MESSAGE_CHAIN.node_pattern_union} $...)
         PATTERN
 
+        def_node_matcher :hash?, '(hash ...)'
+
         def on_send(node)
           message_chain(node) do |(first, *remaining)|
             return if first.to_s.include?('.') || remaining.any?
 
+            return if hash?(first) && first.children.count != 1
```
Comment (dgollahon) [inline lib/rubocop/cop/rspec/single_argument_message_chain.rb]: You could, alternatively, just match a hash with a single pair. I think it might just be `'(hash pair)'`

File: lib/rubocop/cop/rspec/single_argument_message_chain.rb [hunk 5]
```diff
@@ -22,22 +22,42 @@ class SingleArgumentMessageChain < Cop
           (send _ #{Matchers::MESSAGE_CHAIN.node_pattern_union} $...)
         PATTERN
 
+        def_node_matcher :hash?, '(hash ...)'
+
         def on_send(node)
           message_chain(node) do |(first, *remaining)|
             return if first.to_s.include?('.') || remaining.any?
```
Comment (dgollahon) [inline lib/rubocop/cop/rspec/single_argument_message_chain.rb]: I think you could also just match a single argument in the node pattern instead of having to explicitly check if `remaining` is empty.

Comment (dgollahon) [inline lib/rubocop/cop/rspec/single_argument_message_chain.rb]: Though i guess that code wasn't introduced in this PR.

File: spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb [hunk 6]
```diff
@@ -47,6 +47,39 @@
...
+            allow(foo).to receive_message_chain(bar: 42)
+                          ^^^^^^^^^^^^^^^^^^^^^ Use `receive` instead of calling `receive_message_chain` with a single argument.
+          end
+        RUBY
+      end
+
+      include_examples(
+        'autocorrect',
+        'before { allow(foo).to receive_message_chain(bar: 42) }',
+        'before { allow(foo).to receive(:bar) { 42 } }'
```
Comment (dgollahon) [inline spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb]: Do you prefer this syntax to `.and_return(42)`? If so, just curious as to why (i'm not asking you to change anything, just wondering).

File: lib/rubocop/cop/rspec/single_argument_message_chain.rb (hunk 4 above)
Comment (dgollahon) [inline lib/rubocop/cop/rspec/single_argument_message_chain.rb]: You could also capture the key and the value in this expression to make the autocorrect clearer, if you wanted.

File: spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb (hunk 3 above)
Comment (dgollahon) [inline spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb]: If so, it would actually make https://github.com/backus/rubocop-rspec/pull/375/files#diff-91336a4d8c49f0937c8d89c6ea457d1cR58 simpler by just being `#to_s`, i think.

**Comment by Darhazer** [comment]
Thank you for the review @dgollahon 
I'll update it soon

File: spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb (hunk 6 above)
Comment (Darhazer) [inline spec/rubocop/cop/rspec/single_argument_message_chain_spec.rb]: Actually I prefer the `.and_return`. I left the block one just for consistency with the other test cases. On the other hand `and_return` is closer to the original code (no lazy evaluation), so I will replace it with `and_return`

**Comment by Darhazer** [comment]
@dgollahon I couldn't actually apply most of your advice, but made the node matchers to match only single arguments / hash with single key; and also changed the autocorrect to `and_return`

**Comment by dgollahon** [comment]
@Darhazer: Alrighty. I'll take a glance after work 👍

File: lib/rubocop/cop/rspec/single_argument_message_chain.rb [hunk 7]
```diff
@@ -33,11 +37,27 @@ def on_send(node)
...
+          key, value = *arg.children.first
+
+          corrector.replace(arg.loc.expression, key_to_arg(key))
+          corrector.insert_after(arg.parent.loc.end,
+                                 ".and_return(#{value.source})")
+        end
+
+        def key_to_arg(node)
+          key, = *node
+          ":#{key.to_sym}"
```
Comment (dgollahon) [inline lib/rubocop/cop/rspec/single_argument_message_chain.rb]: Pretty sure this `#to_sym` is unnecessary.

What do you think about

`node.sym_type? ? ":#{key}" : node.source`

to preserve the string or symbol type of keys? I think it would be better if the autocorrect were as true to the original as possible and the choice between string and symbol keys could be left to another cop.

Comment (dgollahon) [inline lib/rubocop/cop/rspec/single_argument_message_chain.rb]: At least, I don't think there's any case where `#to_sym` and then an implicit `#to_s` is different than just the implicit `#to_s`.

I've realized the code as it stands actually has a bug though.

Try adding:

```
include_examples(
  'autocorrect',
  'before { allow(foo).to receive_message_chain(:"#{foo}" => 42) }',
  'before { allow(foo).to receive(:"#{foo}").and_return(42) }'
)
```

to the test suite (you can also observe the same issue with an interpolated string rather than a symbol).

My suggestion fixes these issues, but I don't know if there are potentially any other corner cases.

We could also, of course, limit the node pattern to only match against sym, dsym, str, and dstr. The way i have it now seems to handle sends and constants though.

ex:

```
receive_message_chain(foo => 42)
```
receive_message_chain(FOO => 42)
```

I can't think of any other semi-reasonable case at the moment.

**Comment by Darhazer** [comment]
@dgollahon thanks for the help. I've applied `node.sym_type? ? ":#{key}" : node.source` and added the test case you've mentioned
